### PR TITLE
Mehul.sonowal/add tags to pull datasets tests final

### PR DESCRIPTION
**Pull request by mehulsonowal**
## Description

Add support for filtering dataset records by tags when pulling datasets from LLM Observability.

This PR introduces the ability to pass a `tags` parameter to `LLMObs.pull_dataset()`, allowing users to filter dataset records by tag key-value pairs. It also upgrades several API endpoints from `/api/unstable/` to `/api/v2/` for improved stability/ feature support.

### Key Changes

**API Changes (`ddtrace/llmobs/_llmobs.py`):**
- Added optional `tags` parameter to `pull_dataset()` method
- Added input validation to ensure tags is a list of strings

**Writer Updates (`ddtrace/llmobs/_writer.py`):**
- Upgraded `dataset_batch_update` endpoint from `/api/unstable/` to `/api/v2/`
- Upgraded `dataset_get_with_records` endpoints from `/api/unstable/` to `/api/v2/`
- Added `tags` parameter to `dataset_get_with_records()` method
- Added `filter[tags]` query parameter support for record filtering
- Added `tags` field parsing from API responses when fetching records
- Added `tags` field serialization when creating/updating records via `_get_record_json()`

**Dataset Model Updates (`ddtrace/llmobs/_experiment.py`):**
- Added `tags` field to `DatasetRecordRaw` TypedDict
- Added `filter_tags` property to `Dataset` class to track applied tag filters
- Updated `Experiment` to write `filtered_record_tags` to config when dataset has tag filters
- Updated experiment runs to apply record tags as span tags

**New Tests (`tests/llmobs/test_experiments.py`):**
- `test_dataset_pull_with_tags` - Basic tag filtering
- `test_dataset_pull_with_nonexistent_tags` - Filtering with non-matching tags
- `test_dataset_pull_with_partial_tag_match` - Filtering with subset of tags
- `test_dataset_pull_with_one_matching_one_nonexistent_tag` - Mixed tag filtering
- `test_dataset_pull_without_tags_returns_all_records` - No filter returns all
- `test_dataset_pull_with_single_tag` - Single tag filtering
- `test_dataset_pull_with_tags_and_project` - Tag filtering with custom project

**Test Infrastructure (`tests/llmobs/conftest.py`):**
- Added `register_dataset_for_cleanup()` function for session-scoped dataset cleanup
- Added `cleanup_registered_datasets` session-scoped fixture for proper test cleanup

Multiple tag-related tests share the same dataset fixture (`test_dataset_one_record_with_tags`). Each test calls the API with different filter parameters (e.g., `filter[tags]=env:prod` vs `filter[tags]=nonexistent:tag`), which creates different VCR cassette files based on the URL path. If the dataset were deleted after the first test, subsequent tests would fail with 404 errors when VCR attempts to make real API calls for cache misses. The session-scoped cleanup defers dataset deletion until all tests complete, ensuring the dataset remains available for all tag filtering test variants.

## Testing

- VCR cassettes have been regenerated for the v2 endpoints
- Tests validate tag filtering behavior, including edge cases like non-existent tags and partial matches
- Discovered backend API bug when uploading large batches - see https://github.com/DataDog/dd-source/pull/358889

## Risks

- **API Migration**: Endpoints upgraded from `/api/unstable/` to `/api/v2/`. Ensure backend supports v2 endpoints before deployment.
- **Backward Compatibility**: The new `tags` parameter is optional with default `None`, so existing code is unaffected.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -4,7 +4,7 @@ exclude-file = .codespellignorelines
 ignore-words-list = asend,dne,fo,medias,ment,nin,ot,setttings,statics,ba,spawnve,doas
 
 # DEV: We use `conftest.py` as a local pytest plugin to configure hooks for collection
-[tool:pytest]
+# [tool:pytest]
```
Comment (gary-huang) [inline setup.cfg:7]: lets not check this in

File: ddtrace/llmobs/_writer.py [hunk 2]
```diff
@@ -455,13 +460,14 @@ def _get_record_json(record: Union[UpdatableDatasetRecord, DatasetRecordRaw], is
...
         dataset_id: str,
+        project_id: str,
         insert_records: List[DatasetRecordRaw],
         update_records: List[UpdatableDatasetRecord],
         delete_record_ids: List[str],
     ) -> Tuple[int, List[str]]:
         irs: JSONType = [self._get_record_json(r, False) for r in insert_records]
         urs: JSONType = [self._get_record_json(r, True) for r in update_records]
-        path = f"/api/unstable/llm-obs/v1/datasets/{dataset_id}/batch_update"
+        path = f"/api/v2/llm-obs/v1/{project_id}/datasets/{dataset_id}/batch_update"
```
Comment (gary-huang) [inline ddtrace/llmobs/_writer.py:470]: just curious, do we have to migrate these endpoints this time around? or are you just getting them out of the way in one go?

File: ddtrace/llmobs/_writer.py [hunk 3]
```diff
@@ -441,10 +441,15 @@ def _get_record_json(record: Union[UpdatableDatasetRecord, DatasetRecordRaw], is
...
 
+        tags: JSONType = None
+        if "tags" in record:
+            tags = [] if record["tags"] is None else record["tags"]
+
         rj: JSONType = {
             "input": cast(Dict[str, JSONType], record.get("input_data")),
             "expected_output": expected_output,
             "metadata": metadata,
+            "tags": tags,
```
Comment (gary-huang) [inline ddtrace/llmobs/_writer.py:452]: this field is only supported in `insert_records` of the request, i think this is ignored for `update_records`

i havent dug deeper into the impl yet but i dont think we support updating tags for a given record from the SDK, im inclined to say lets leave it out and indicate that it's unsupported, but lets check in with product on this

File: ddtrace/llmobs/_writer.py (hunk 2 above)
Comment (mehulsonowal) [reply ddtrace/llmobs/_writer.py:470]: I believe this was necessary for the tests since I needed to upload records with tags attached

Comment (gary-huang) [reply ddtrace/llmobs/_writer.py:470]: ah ok gotcha
